p2p/transport/mem/memoryconn.go

```diff
@@ -0,0 +1,156 @@
+package memorytransport
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"runtime"
+
+	"github.com/glycerine/rbuf"
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+var InitialBufferSize = 5 * 1024 * 1024
+var BufferGrowthFactor = 2
+
+// GrowableRingBuffer wraps an rbuf.AtomicFixedSizeRingBuf with a mutex,
+// allowing us to grow its capacity should it become filled.
+type GrowableRingBuffer struct {
+	mbuf *sync.RWMutex
+	buf  *rbuf.AtomicFixedSizeRingBuf
+}
+
+func NewGrowableRingBuffer() *GrowableRingBuffer {
+	return &GrowableRingBuffer{
+		mbuf: new(sync.RWMutex),
+		buf:  rbuf.NewAtomicFixedSizeRingBuf(InitialBufferSize),
+	}
+}
+
+func (rb *GrowableRingBuffer) Readable() int {
+	rb.mbuf.RLock()
+	defer rb.mbuf.RUnlock()
+
+	return rb.buf.Readable()
+}
+
+func (rb *GrowableRingBuffer) Write(bytes []byte) (int, error) {
+	rb.mbuf.RLock()
+	written, err := rb.buf.Write(bytes)
+	if err != nil {
+		// Buffer was short, grow
+		rb.mbuf.RUnlock()
+		rb.mbuf.Lock()
+		defer rb.mbuf.Unlock()
+		newSize := rb.buf.N * BufferGrowthFactor
+		oldBuf := rb.buf
+		rb.buf = rbuf.NewAtomicFixedSizeRingBuf(newSize)
+		rb.buf.Adopt(oldBuf.Bytes(false))
+		return rb.buf.Write(bytes[written:])
+	}
+	rb.mbuf.RUnlock()
+	return written, nil
+}
+
+func (rb *GrowableRingBuffer) Read(out []byte) (int, error) {
+	rb.mbuf.RLock()
+	defer rb.mbuf.RUnlock()
+
+	return rb.buf.Read(out)
+}
+
+// MemoryConn is a view into a pair of GrowableRingBuffers representing a fully
+// duplexed connection. One is used for reading and the other is used for
+// writing.
+type MemoryConn struct {
+	readbuf  *GrowableRingBuffer
+	writebuf *GrowableRingBuffer
+	parent   *MemoryConnPair
+}
+
+func NewMemoryConn(reader, writer *GrowableRingBuffer, parent *MemoryConnPair) *MemoryConn {
```

Comment:
We may not need any buffering at all. In that case, we can implement this entirely with channels:

```go
func (c *ChanConn) Write(b []byte) (int, error) {
  n := 0
  for n < len(b) {
    select {
    case <-c.writeTimer.C:
      return n, ErrTimeout
    case <-c.closed:
      ...
    case remote, ok := <-c.write:
      if !ok { ... }
      coppied := copy(remote, b[:n])
      c.writeResult <- coppied
      n += coppied
    }
  }
}

func (c *ChanConn) Read(b []byte) (int, error) {
  select {
    case <-c.readTimer.C:
      return 0, ErrTimeout
    case <-c.closed:
      ...
    case c.read <- b:
  }
  return <-c.readResult, nil
}


// avoids any intermediate buffers in `io.Copy`.
func (c *ChanConn) ReadFrom(r io.Reader) (int64, error) {
  n := 0
  for {
    select {
    case <-c.writeTimer.C:
      return n, ErrTimeout
    case <-c.closed:
      ...
    case remote, ok := <-c.write:
      if !ok { ... }
      coppied, err := r.Read(remote)
      c.writeResult <- coppied
      n += coppied
      if err != nil {
        return n, err
      }
    }
  }
}


// c.write is connected to c.read, c.writeResult is connected to c.readResult.
```

That should use very little memory and should avoid copying when at all possible.

Reply:
Actually, it looks like `net.Pipe` does everything except `ReadFrom` (which isn't absolutely necessary. We should probably just use that.

Reply:
yeahhhhhh i was thinking about this too! i'm happy to just keep the pipe.